### DISCUSSION TITLE
Loading and saving to a sqlite.db file

### DISCUSSION THREAD

**Discussion by jlarmstrongiv** [rhashimoto/wa-sqlite#82] [Ideas]
I would like to add a `save` and `load` database functions to `@vlcn.io/crsqlite-wasm`.  

I found that package [depends on](https://github.com/vlcn-io/cr-sqlite/blob/main/js/packages/crsqlite-wasm/src/index.ts#L12) `wa-sqlite`, and I think the addition would be best suited for it.

I found a [previous discussion](https://github.com/rhashimoto/wa-sqlite/discussions/72#discussioncomment-4620789) which describes:

> There is currently no example code in this repo for this and I have no plans to add it. I would probably approach it by instantiating a VFS and calling its methods - xOpen, xLock, xFileSize, xSectorSize, xRead, xUnlock, xClose - to read data in sector sized chunks...but that requires an understanding of both the [SQLite VFS interface](https://www.sqlite.org/c3ref/vfs.html) and [how wa-sqlite adapts it](https://github.com/rhashimoto/wa-sqlite/blob/master/src/VFS.js) to Javascript. A less general but probably easier way to do it would be to copy the [code that opens the IndexedDB database](https://github.com/rhashimoto/wa-sqlite/blob/1036b7f40e46d571b9489a6a0394a0833b53f46c/src/examples/IDBBatchAtomicVFS.js#L628-L644) and then read the blocks directly with the IndexedDB API.

I also found that `sql.js` supports loading a database via:
```js
/* loading */
dbFileElm.onchange = () => {
  const f = dbFileElm.files[0];
  const r = new FileReader();
  r.onload = function() {
    const Uints = new Uint8Array(r.result);
    db = new SQL.Database(Uints);
  }
  r.readAsArrayBuffer(f);
}

const sqlPromise = initSqlJs({
  locateFile: file => `https://path/to/your/dist/folder/dist/${file}`
});
const dataPromise = fetch("/path/to/database.sqlite").then(res => res.arrayBuffer());
const [SQL, buf] = await Promise.all([sqlPromise, dataPromise])
const db = new SQL.Database(new Uint8Array(buf));

/* saving */
const fs = require("fs");
// [...] (create the database)
const data = db.export();
const buffer = Buffer.from(data);
fs.writeFileSync("filename.sqlite", buffer);
```

Would you be open to adding `save` and `load` functions to your package?  Do you have suggestions on the best way for me to implement those?

**Comment by rhashimoto** [comment]
I guess I do have a few thoughts and suggestions...

As you excavated from the other thread, I would instantiate a VFS and make calls to its methods defined in the [VFS base class](https://github.com/rhashimoto/wa-sqlite/blob/master/src/VFS.js). It might be instructive to uncomment the browser debug console logging in the VFS that you use, e.g. [here](https://github.com/rhashimoto/wa-sqlite/blob/035a7cdfe4df388d28453fdadb01bd1f614e3264/src/examples/IDBBatchAtomicVFS.js#L23) for IDBBatchAtomicVFS, and open your application or the demo to watch SQLite make calls because you'll need to mimic that to some degree. I _think_ you can do what you need with only `xOpen`, `xRead`, `xWrite`, and `xClose`, but I'm not 100% sure across all the VFS classes.

One critically important thing to know is the IndexedDB VFS classes store each write as a separate IndexedDB object, and the VFS assumes that 1 object = 1 page in a database file (except during a VACUUM which you won't encounter) and reads will never cross a page boundary, because that's what SQLite does (and there are [advantages](https://github.com/rhashimoto/wa-sqlite/discussions/46) for that). That's a key constraint on how you perform I/O with the VFS. In addition, page size is not a known constant value (it can be changed with SQL) so you'll need to read the [page size](https://www.sqlite.org/fileformat.html#page_size) from the [file header](https://www.sqlite.org/fileformat.html#the_database_header) first.

Also the API for VFS classes is changing so you'll almost certainly want to write to the [new API](https://github.com/rhashimoto/wa-sqlite/discussions/76).

I would recommend that the interface to your code stream file data if possible (e.g. with the [Streams API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API)), to allow for the possibility of managing databases larger than available memory. I don't know how well this is supported yet, though - I do know you can't use fetch to upload a stream.

You're welcome to develop this in your own repo. As for adding it to this project, it might make sense to have this in examples/ and a demo page in demos/. Integration with the main demo might be nice, too, but be aware it has been rewritten in the same branch with the new APIs. This would be the first substantive code submission, so you might have to bear with me while I figure out how I want to manage PRs, could be a little annoying for you as I make things up on the fly. 🤪 (but not entirely kidding)

**Reply by jlarmstrongiv** [reply]
@rhashimoto thank you for the thoughts and suggestions! 😄

I will enable logging with `IDBBatchAtomicVFS` and try that out first.  My goal is to get it working in the browser and that’s a great place to start. 

I’m familiar with byte parsing with [restructure](https://www.npmjs.com/package/restructure), so I’ll be able to jump right in to get the page size from the header.

> Also the API for VFS classes is changing so you'll almost certainly want to write to the https://github.com/rhashimoto/wa-sqlite/discussions/76.

Would you make an alpha release so that I can try the latest API?  Definitely would like to work with the `Uint8Array`s

I’m familiar with the [web streams apis](https://exploringjs.com/nodejs-shell-scripting/ch_web-streams.html), and I’ll definitely need to use those to support larger databases like you said.

Once I have something working for both reading and writing to SQLite files in the browser, I’ll work with you to add it to the repo 👍

**Reply by rhashimoto** [reply]
> Would you make an alpha release so that I can try the latest API? Definitely would like to work with the Uint8Arrays

You can develop against the [breaking-changes branch](https://github.com/rhashimoto/wa-sqlite/tree/breaking-changes) until it gets merged to the trunk. The changes I'm still making in the branch shouldn't affect you.

> I’m familiar with byte parsing with [restructure](https://www.npmjs.com/package/restructure), so I’ll be able to jump right in to get the page size from the header.

One warning - getting me to accept a new package dependency is a really high bar if you eventually want to submit a PR. In this specific case of extracting a couple fields from the header, you should just use [DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView).

**Reply by jlarmstrongiv** [reply]
> You can develop against the [breaking-changes branch](https://github.com/rhashimoto/wa-sqlite/tree/breaking-changes) until it gets merged to the trunk. The changes I'm still making in the branch shouldn't affect you.

Alrighty, I’m currently installing the dependencies to build it locally

> One warning - getting me to accept a new package dependency is a really high bar if you eventually want to submit a PR. In this specific case of extracting a couple fields from the header, you should just use [DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView).

DataView it is!  Thanks for the heads up

**Reply by rhashimoto** [reply]
The breaking-changes branch has pre-built WASM artifacts in dist/, so you shouldn't need to do an Emscripten build for this unless you want to.

**Reply by jlarmstrongiv** [reply]
Fantastic 🙌 that’s exactly what I was looking for.  Thank you!

**Comment by sgbeal** [comment]
>  I think you can do what you need with only xOpen, xRead, xWrite, and xClose, but I'm not 100% sure across all the VFS classes.

A potential gotcha: _some_ VFSes require that the lengths given to xWrite and xRead be a legal db page size (a power of 2 between 512 and 64k). The library internals only ever pass such sizes to those functions. Not all VFSes are picky about that, though.

**Reply by rhashimoto** [reply]
Yes, that's what I was trying to get across but perhaps it wasn't clear. The wa-sqlite IndexedDB VFS classes expect 1 write = 1 IndexedDB object = 1 page, not just a legal page size but the actual page size. Reads don't have to be a full page but can't straddle a page boundary.

**Reply by jlarmstrongiv** [reply]
Thank you both!  That‘s a great tip to be aware of

**Comment by tantaman** [comment]
To summarize and ensure I understand, the basic idea for importing a db file is to:

1. Read the db header (defined here: https://www.sqlite.org/fileformat.html)
2. Pull the page size from that header
3. Pull the page count from the header
4. Starting at byte index 100, read `PAGE_SIZE` chunks up until `PAGE_COUNT`
5. For each chunk, write it to the vfs via `xWrite`

**Reply by rhashimoto** [reply]
You'll need to surround your file accesses with `xOpen` and `xClose`.

Pages start at the beginning of the file so (4) is not correct. The header is part of the first page.

**Reply by jlarmstrongiv** [reply]
This big-picture overview and tip helped a lot!  For writing to a db file, I assume the steps are:
- reading every page for the given DB
- concatenate/stream all the bytes together

**Comment by sgbeal** [comment]
MAINTAINER UPDATE 4/18/23: I'm hiding this comment because isn't true for the VFS classes we're discussing.

> Pull the page size from that header

You don't strictly need to do that. Far simpler is to figure out a _compatible_ block size by looking at the full size of the db. The db size is always a multiple of 512, so you can safely just read/write blocks of that size. If you really want to reduce the number of reads/writes, you can instead calculate a best-fit block size. That is, find the first value of the following list which is also a multiple of the db file's size:

- 65536
- 32768
- 16384
- 8192
- 4096
- 2048
- 1024
- 512

In a pinch, though, 512 will always be okay.

The xRead/xWrite APIs don't actually know which DB they're working on, so don't know its page size. _Some_ of them internally require that the size passed to them is a legal page size, though.

**Reply by rhashimoto** [reply]
Again:

> The wa-sqlite IndexedDB VFS classes expect 1 write = 1 IndexedDB object = 1 page, not just a legal page size but the actual page size.

**Comment by AntonOfTheWoods** [comment]
```
  const res = await fetch("http://localhost/mytest.db");
  if (!res.ok) throw new Error("bad response");

  const buffer = new Uint8Array(await res.arrayBuffer());
  const pOut = new Value();
  const vfs = new AccessHandlePoolVFS('/a_directory');
  await vfs.isReady;
  const FILE_ID = 0xdeadbeef;
  const filename = 'a_filename';
  vfs.xOpen(filename, FILE_ID, VFS.SQLITE_OPEN_CREATE | VFS.SQLITE_OPEN_READWRITE  | VFS.SQLITE_OPEN_MAIN_DB, pOut.pass());
  vfs.xWrite(FILE_ID, buffer, 0);
  vfs.xSync(FILE_ID);
  vfs.xClose(FILE_ID);
  await vfs.close();

```
The above looks suspiciously like it works... I get the contents of the tables I'm expecting and a `PRAGMA integrity_check;` returns `'ok'`. I got inspiration from the tests - I have no idea what the `pOut.pass()` is doing... Is it doing something I should know about? If my (currently anyway) only use case is importing files like the above - is there anything else I am missing? If I were to have a list of chunks of binary file, is there anything stopping me from just writing them in chunks (in order of course!)?

EDIT: Added the missing `VFS.SQLITE_OPEN_MAIN_DB` flag, otherwise the imported db gets deleted on worker reload...

**Reply by rhashimoto** [reply]
I think that will work for AccessHandlePoolVFS. IDBBatchAtomicVFS is a lot more temperamental and most of its methods are asynchronous; your code won't work as-is with it.

xOpen returns flags using pOut which you won't care about. You just need to pass a valid DataView in, which is what the tests are doing.

**Reply by AntonOfTheWoods** [reply]
Awesome, thanks, `AccessHandlePoolVFS` is now working very well. For `IDBBatchAtomicVFS`, I tried sticking an `await` in front of the methods and only `xWrite`ing in blocks of `page_size` and it *looks* like I can write everything... until I try and open the database after the initial install, and then get `TypeError: d.data.subarray is not a function` coming from `buffer.set(block.data.subarray(blockOffset, blockOffset + nBytesToCopy));`. Any pointers most appreciated!

**Reply by rhashimoto** [reply]
Not enough information for remote debugging. Can you show me what the IndexedDB object in question looks like, e.g. in Chrome DevTools?

![Screen Shot 2023-08-15 at 7 13 53 AM](https://github.com/rhashimoto/wa-sqlite/assets/156154/eab955e9-baf0-4373-9ba6-acd3c9f6a44d)

**Reply by rhashimoto** [reply]
My guess is you're passing a DataView to xWrite because that's what your post above does. It should be a Uint8Array instead.

**Reply by AntonOfTheWoods** [reply]
Thanks again. I am a little further along - I was indeed using a `DataView` but I am still unable to get a valid file saved. Here is what I am using:

```javascript
  const vfs = new IDBBatchAtomicVFS(`/tcdb`);
  const res = await fetch("http://localhost/my.db");
  if (!res.ok) throw new Error("Could not fetch the database");
  let result = await vfs.xOpen(
    filename,
    FILE_ID,
    VFS.SQLITE_OPEN_CREATE | VFS.SQLITE_OPEN_READWRITE | VFS.SQLITE_OPEN_MAIN_DB,
    new DataView(new ArrayBuffer(8)),
  );
  const ab = await res.arrayBuffer();
  if (ab.byteLength % PAGE_SIZE !== 0) throw new Error("The file is not a multiple of the page size");
  for (let i = 0; i < ab.byteLength; i += PAGE_SIZE) {
    result += await vfs.xWrite(FILE_ID, new Uint8Array(ab.slice(i, i + PAGE_SIZE)), i);
  }
  await vfs.xSync(FILE_ID, 0);
  await vfs.xClose(FILE_ID);
  await vfs.close();
  console.log("And the total of the results is", result);
```

The above gives me something very similar to your screenshot:

![image](https://github.com/rhashimoto/wa-sqlite/assets/2504480/d0c760b8-655f-48cb-8fa8-6cc34a5401e2)

And I can `pragma page_size;` which gives me the expected `32768` but then any query or `integrity_check` gives me `Error: database disk image is malformed`. The value of `result` is also zero, so things appear to be saving ok.

**Reply by rhashimoto** [reply]
Are there any warnings or errors in the Javascript console during the database creation and loading?

Show me the first 32 or so bytes of the first page (offset 0). You can search for path ["/tcdb", 0] (I think double quotes are required).

**Reply by AntonOfTheWoods** [reply]
Nevermind... I can't work out what I was doing before but the above code actually does work... it is very slightly different to what I had (I cleaned up to put here) and this cleaned up version actually does produce a valid db. Sigh...

Thanks!

**Reply by AntonOfTheWoods** [reply]
**BUT BUT BUT!!!!!!!!!!!!!!!!!!!**

You MUST NOT have a significant amount of time between the `xWrite`s. I'm picking it's got to do with the flushing but if you don't basically get the entire database and do everything in a single loop then the above WILL NOT WORK. I was getting chunks of my db from the network and `xWrite`ing them as they arrived and this doesn't work, at least not with the 31 10MB blocks I was trying to do it with. I spent several hours trying to work out what was going on - I got down to the fact that md5s of the chunks were identical with getting the entire file and splitting, then doing in chunks... the only thing different between downloading chunks from downloading the entire file and splitting was the timing... And adding a short `sleep` in between iterations proved you can definitely produce a corrupt DB by taking more than a couple of seconds (3000ms?) to import the entire file from the first `xWrite` to the last.

EDIT: It appears that you also MUST have a sufficient amount of time after the `xOpen` (or maybe VFS object creation) before the writes start, possibly also around 3000ms... And even then, it appears to quite regularly create corrupt dbs... 

Is there a way to guarantee it does what it should, or should I just go with SQL inserts?

**Reply by rhashimoto** [reply]
> And adding a short sleep in between iterations proved you can definitely produce a corrupt DB by taking more than a couple of seconds (**3000ms?**) to import the entire file from the first xWrite to the last.

Oh, I see (I think). Chrome (at least, possibly other browsers, too) automatically aborts an IndexedDB transaction after a minute or so. IDBBatchAtomicVFS.xWrite is usually synchronous, but if it's _always_ synchronous then a long SQLite transaction will cause an IndexedDB transaction timeout. To work around that there's [this code](https://github.com/rhashimoto/wa-sqlite/blob/122c46ea4d2f20308c0a3d62b79add048a2bef9c/src/examples/IDBBatchAtomicVFS.js#L214-L229) which switches from synchronous to asynchronous every...3 seconds (bing!). But this is complicated because of the way Asyncify makes two calls for each asynchronous call. You're not calling via Asyncify so you only make one call, and I thought the code would [handle that](https://github.com/rhashimoto/wa-sqlite/blob/122c46ea4d2f20308c0a3d62b79add048a2bef9c/src/examples/IDBBatchAtomicVFS.js#L217) but it certainly sounds like it doesn't.

I'll take a deeper look but it might not be all that soon. You could try to patch IDBBatchAtomicVFS and make [#xWriteHelper](https://github.com/rhashimoto/wa-sqlite/blob/122c46ea4d2f20308c0a3d62b79add048a2bef9c/src/examples/IDBBatchAtomicVFS.js#L239) a public method and call that instead of xWrite. If you do that and your database loads can take a significant fraction of a minute, then you should also make Javascript start a new task every once in a while so the current IndexedDB transaction can commit.

The required delay after xOpen is still a mystery to me, not sure what would explain that.

**Reply by AntonOfTheWoods** [reply]
Thanks again for the help. 

Ok, so I kept looking and it actually doesn't appear to be strictly related to those flushes but rather `MAX_TRANSACTION_LIFETIME_MILLIS` - so 5000ms. If there is not 5000ms between the `xOpen` and the first `xWrite` (or less if I reduce that const - at 2000ms it is also 100% reproducible), then it fails. I tried pretty much everything - it looked like I could just force a 

```
  async forceCommit() {
    this.#tx?.commit();
    await new Promise((resolve) => setTimeout(resolve));
    this.#tx = null;
  }

```

after the `xOpen` in `IDBContext`, but that makes no difference whatsoever... So there is obviously something else that is necessary...

Interestingly, in the actual `xWrite` loop, 2-3 `await new Promise((resolve) => setTimeout(resolve, 0))` will cause db corruption but if there are none (or anything else that allows the next task to execute), or just 1-2 `await new Promise((resolve) => setTimeout(resolve, 0))`, then it executes fine. This happens both whether `MAX_TRANSACTION_LIFETIME_MILLIS` is 5000ms or 2000ms, so I'm not sure what's going on.

So basically, you MUST wait `MAX_TRANSACTION_LIFETIME_MILLIS` after `xOpen` and you MUST NOT let another transaction happen between the start of the `xWrite` and the `vfs.xSync(FILE_ID, 0)`/`vfs.xClose(FILE_ID)`. If both these are true, it now seems pretty reliable that a valid db is produced.

**Reply by rhashimoto** [reply]
Thanks for documenting all that behavior! It's unexpected and strange. I'll see if I can reproduce your findings and determine what's different from how SQLite makes these calls.

https://github.com/rhashimoto/wa-sqlite/issues/110

**Comment by tantaman** [comment]
@AntonOfTheWoods - did you end up writing an exporter to export the DB File from the IndexedDBBatchAtomic VFS?

**Reply by AntonOfTheWoods** [reply]
I didn't @tantaman , and it's not on the radar either, as for me that VFS is only used in a chrome ext, and users will always have the website where they will be able to download from the same DB using AHP (or at least they will when that gets implemented!).

**Reply by rhashimoto** [reply]
If you're building your own SQLite (as @tantaman does), on any platform (not just for the browser), one way to enable database export (but not import) is with the [SQLITE_DBPAGE](https://www.sqlite.org/dbpage.html) virtual table. This module is bundled in the SQLite source and is activated with a compile-time flag. If included, you can fetch raw database file pages using SQL to avoid messing with the VFS directly.

**Comment by rhashimoto** [comment]
I created a demo page, [online here](https://rhashimoto.github.io/wa-sqlite/demo/file/) ([source](https://github.com/rhashimoto/wa-sqlite/tree/master/demo/file)) to demonstrate database file import (local file to wa-sqlite) and export (wa-sqlite to local file). The demo uses direct calls to IDBBatchAtomicVFS, as suggested earlier in this thread. I think it should work with other VFS classes, but I have not tried any.

I built this to investigate Issue #110 (now resolved), based on racy behavior mentioned earlier in this thread. ~I have not reproduced any of that behavior, but that bug remains open. It is possible that my code doesn't produce those issues and all is fine, or that it will and I just haven't seen it yet.~ I did try uploading and downloading database files between 12 KB and 6.3 GB, and transfers in both directions pass `PRAGMA integrity_check`.

The implementation streams data using ReadableStream in both import and export, which allows handling database files larger than available memory. Note that the only cross-browser way (that I know of) to stream an export to a local file is with a service worker, so that's what I did. I wrote my service worker as an ES6 module, so it would need to be bundled to work on Firefox (which doesn't yet support ES6 module service workers).

**Reply by AntonOfTheWoods** [reply]
Thanks very much for this. I am absolutely snowed under and have a new project and have had to move back from Linux to Windows for my current project, so will need a few days to see whether I can reproduce. I was very reliably able to reproduce, so I'll try and do that again so we can see where the difference is!

**Reply by rhashimoto** [reply]
Sounds good. Follow up with anything to do with racy behavior over in #110.

**Reply by rhashimoto** [reply]
> I think it should work with other VFS classes, but I have not tried any.

Since the demo exports using a service worker, drop-in replacement of course won't work with a VFS that uses OPFS synchronous access handles (i.e. _every_ OPFS VFS) because those aren't available in a service worker. It shouldn't be difficult to work around this with the `File` interface instead, though, just need a way to get the FileSystemFileHandle. Alternatively, you could proxy the VFS method calls to a dedicated Worker but that sounds harder.

> Note that the only cross-browser way (that I know of) to stream an export to a local file is with a service worker, so that's what I did.

You _can_ use [URL.createObjectURL()](https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL_static) to stream a `File` to a local file, so that would not require a service worker for a database file in OPFS. However, that won't work for an AccessHandlePoolVFS database, because that implementation prepends a sector of VFS metadata to the database file. In retrospect, that was an unfortunate design decision (one that the official SQLite WASM project decided to copy despite an [alternative solution](https://github.com/rhashimoto/wa-sqlite/discussions/99)) so the service worker approach would be needed there as well.

URL.createObjectURL() can also take a MediaSource argument so I wondered whether creating a custom MediaSource would work for AccessHandlePoolVFS, but it sounds like this is [now deprecated](https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL_static#using_object_urls_for_media_streams).
